src/phase-management/PhaseAdvancer.js

```diff
@@ -0,0 +1,249 @@
+const config = require("config");
+const { Engine } = require("json-rules-engine");
+
+const rulesJSON = require("./phase-rules.json");
+const errors = require("../common/errors");
+
+const helper = require("../common/helper");
+
+// Helper functions
+
+// TODO: Move these to a common place
+
+const normalizeName = (name) => name.replace(/ /g, "");
+
+const shouldCheckConstraint = (operation, phase, constraintName, rules) => {
+  const normalizedConstraintName = normalizeName(constraintName);
+  return (
+    operation === "close" &&
+    phase.constraints &&
+    rules.constraintRules[phase.name]?.includes(normalizedConstraintName)
+  );
+};
+
+const parseDate = (dateString) => {
+  const date = new Date(dateString).getTime();
+  return isNaN(date) ? null : date;
+};
+
+// End of helper functions
+
+class PhaseAdvancer {
+  #rules = rulesJSON;
+
+  #factGenerators = {
+    Registration: async (challengeId) => ({
+      registrantCount: await this.#getRegistrantCount(challengeId),
+    }),
+    Submission: async (challengeId) => ({
+      submissionCount: await this.#getSubmissionCount(challengeId),
+      hasActiveUnreviewedSubmissions: await this.#hasActiveUnreviewedSubmissions(challengeId),
+    }),
+    Review: async (challengeId) => ({
+      allSubmissionsReviewed: await this.#areAllSubmissionsReviewed(challengeId),
```

Comment:
There is another fact that checks if all test cases are uploaded If there is a development reviewer for challenge. But this could be something we are not doing anymore.